### PR TITLE
gpcheckcat: add check for orphaned TOAST tables

### PR DESCRIPTION
**Pull request by kalensk**
During the course of this PR, several refactors were made for code hygiene and to help implement the feature.

There are four main types of orphan TOAST tables:
- Double Orphan TOAST tables due to a missing reltoastrelid in pg_class and a missing pg_depend entry.
- Bad Reference Orphaned TOAST tables due to a missing reltoastrelid in pg_class.
- Bad Dependency Orphaned TOAST tables due to a missing entry in pg_depend.
- Mismatch Orphaned TOAST tables due to reltoastrelid in pg_class pointing to an incorrect TOAST table.

Repair scripts are done on a per-segment basis since catalog changes are required.

Note: A repair script cannot be generated for mismatch orphan TOAST tables 
because the current repair implementation does not work with two or more tables, 
so let the user repair manually for safety. A manual catalogue change is needed 
to fix by updating the pg_depend TOAST table entry and setting the refobjid 
field to the correct dependent table.

### CODE REVIEW

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 1]
```diff
@@ -2600,3 +2600,88 @@ def impl(context, database):
...
+@given('the database "{dbname}" is broken with "{broken}" orphaned toast tables')
+def impl(context, dbname, broken):
+    break_orphaned_toast_tables(context, dbname, broken)
+
+@given('in the database "{dbname}" the toast table for "{table}" is renamed')
+def impl(context, dbname, table):
+    # Normally we would connect with allowSystemTableMods=True. We run into a
+    # bug(?) where the allow_system_table_mods GUC isn't forwarded to the
+    # segments, however. So SET it explicitly inside the SQL session.
+    with dbconn.connect(dbconn.DbURL(dbname=dbname)) as conn:
```
Comment (berlin-ab) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: It might be nice to pull this behavior out into a function so if someone needs to do something similar, it could be reused in python-land.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                elif issue.type == 'reference_orphans':
+                    repair_statement = "UPDATE \"pg_class\" SET reltoastrelid = %d WHERE oid = %s;" % (
+                    row["toast_table_oid"], row["dependent_table_oid"])
+                    content_id_to_segment_map[row['content_id']]['repair_statements'].append(repair_statement)
+                elif issue.type == 'dependency_orphans':
+                    # 1259 is the reserved oid for pg_class and 'i' means internal dependency; these are safe to hard-code
+                    repair_statement = "INSERT INTO pg_depend VALUES (1259, %d, 0, 1259, %d, 0, 'i');" % (
+                    row["toast_table_oid"], row["expected_table_oid"])
+                    content_id_to_segment_map[row['content_id']]['repair_statements'].append(repair_statement)
+                # elif self.issues_to_rows[issue] and issue.type == 'mismatch_orphans':  # The following repair does not work with two or more tables, so let the user repair manually for safety.
```
Comment (berlin-ab) [inline gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: We probably don't need to hold on to these lines of code.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            elif self.issues_to_rows[issue] and issue.type == 'mismatch_orphans':
+                log_output.append(issue.header)
+                log_output.append('''  A manual catalog change is needed to fix by updating the pg_depend TOAST table entry and setting the refobjid field to the correct dependent table.\n''')
+
+        return '\n'.join(log_output)
+
+    def get_repair_statements(self, segments):
+        content_id_to_segment_map = self.__get_content_id_to_segment_map(segments)
+
+        for issue in self.issues_to_rows.keys():
```
Comment (berlin-ab) [inline gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Can we simplify this nested structure?

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 1 above)
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: @berlin-ab Do you mean tweak this current behave function to be more generic such that it can rename any table including a system catalog table?

If so, we would still need a way to query for the rel_oid of the toast table to know what table to rename and pass that along. Something like adding the following two functions? Is this preferred?

```python
@given('in the database "{dbname}" the toast table for "{table}" is renamed')
def impl(context, dbname, table):
    with dbconn.connect(dbconn.DbURL(dbname=dbname)) as conn:
        rel_oid = dbconn.execSQLForSingleton(conn, '''SELECT oid FROM pg_class WHERE relname = '%s';''' % table)
    renamed_table = 'pg_toast.pg_toast_' + rel_oid
    context.execute_steps(u'''
                        And in the database "{dbname}" the "{table}" is renamed to "{renamed_table}"
                        ''')

@given('in the database "{dbname}" the "{table}" is renamed to "{renamed_table}"')
def impl(context, dbname, table):
    # To rename catalog tables we would normally connect with allowSystemTableMods=True. 
    # However, we run into a bug(?) where the allow_system_table_mods GUC isn't
    # forwarded to the segments. So SET it explicitly inside the SQL session.
    with dbconn.connect(dbconn.DbURL(dbname=dbname)) as conn:
        dbconn.execSQL(conn, '''SET allow_system_table_mods=true; ALTER TABLE %s RENAME TO %s''' % table, renamed_table)
        conn.commit()
```

File: gpMgmt/bin/gpcheckcat [hunk 4]
```diff
@@ -2905,6 +2909,51 @@ def log_unique_index_violations(violations):
     logger.error('\n'.join(log_output))
 
 
+# -------------------------------------------------------------------------------
+def checkOrphanedToastTables():
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat]: Personally, I don't find the function size to be problematic. Any specific suggestions? 

All I can think of is extracting out the logging part into a separate function underneath it. Doing so doesn't seem like much value add, but Im open to the idea. Something like the following shown below?

Another idea which does not seem practical with the structure of the current code is encapsulating the logging stuff into the `OrphanedToastTablesCheck` class. However, this would require potentially making the following dependencies of that class: logger, CatOrphanToastTableIssue, and getGPObject - which seems problematic.


```python
def checkOrphanedToastTables():
    logger.info('-----------------------------------')
    logger.info('Performing check: checking for orphaned TOAST tables')

    db_connection = connect2(GV.cfg[1], utilityMode=False)
    checker = OrphanedToastTablesCheck()
    check_passed = checker.runCheck(db_connection)

    checkname = 'orphaned toast table(s)'
    if check_passed:
        logger.info('[OK]  %s' % checkname)
    else:
        logger.info('[FAIL]  %s' % checkname)
        GV.checkStatus = False
        setError(ERROR_REMOVE)

        logOrphanedToastTableIssues()

        do_repair_for_segments(segments_to_repair_statements=checker.get_repair_statements(GV.cfg), issue_type="orphaned_toast_tables", description='Repairing orphaned TOAST tables',)

def logOrphanedToastTableIssues():
    log_output = ['Orphan toast tables detected for the following scenarios:']
    for issue, rows in checker.get_issue_to_rows().iteritems():
        log_output.append('\n' + issue.header)
        log_output.append('content_id | toast_table_oid | toast_table_name | expected_table_oid | expected_table_name | dependent_table_oid | dependent_table_name')
        log_output.append('-----------+-----------------+------------------+--------------------+---------------------+---------------------+---------------------')
        for row in rows:
            log_output.append('{} | {} | {} | {} | {} | {} | {}'.format(
                row['content_id'],
                row['toast_table_oid'],
                row['toast_table_name'],
                row['expected_table_oid'],
                row['expected_table_name'],
                row['dependent_table_oid'],
                row['dependent_table_name']))
    log_output = '\n'.join(log_output)
    log_output = log_output + '\n'
    logger.error(log_output)

    gplog.log_literal(logger, logging.CRITICAL, checker.get_fix_text())

    for table, issue_to_causes in checker.get_table_to_issues().iteritems():
        cat_issue_obj = CatOrphanToastTableIssue(table.oid, table.catname, issue_to_causes)
        error_object = getGPObject(table.oid, table.catname)
        error_object.addOrphanToastTableIssue(cat_issue_obj)
```

File: gpMgmt/bin/gpcheckcat_modules/repair.py [hunk 5]
```diff
@@ -25,26 +25,36 @@ def __init__(self, context=None, issue_type=None, desc=None):
...
-
+    def create_repair_for_extra_missing(self, catalog_table_obj, issues, pk_name, segments):
         catalog_name = catalog_table_obj.getTableName()
         extra_missing_repair_obj = RepairMissingExtraneous(catalog_table_obj=catalog_table_obj,
                                                            issues=issues,
                                                            pk_name=pk_name)
         repair_dir = self.create_repair_dir()
-        all_seg_ids = self._context.report_cfg.keys()
-        segment_to_oid_map = extra_missing_repair_obj.get_segment_to_oid_mapping(all_seg_ids)
+        segment_to_oids_map = extra_missing_repair_obj.get_segment_to_oid_mapping(map(lambda config: config['content'], segments.values()))
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: A huge +1 to that! That was a big issue when in the middle of creating this PR, and seemed out of scope and time for this particular change.

File: gpMgmt/bin/gpcheckcat_modules/repair.py [hunk 6]
```diff
@@ -25,26 +25,36 @@ def __init__(self, context=None, issue_type=None, desc=None):
     def create_repair(self, sql_repair_contents):
         repair_dir = self.create_repair_dir()
 
-        sql_file_name = self.__create_sql_file_in_repair_dir(repair_dir, sql_repair_contents)
+        master_segment, = filter(lambda config: config['content'] == -1, self._context.cfg.values())
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: Taking a closer look, it appears you are correct in that it returns both the master and the master standby.

GV.cfg gets populated from getGPConfiguration() which calls this specific [query](https://github.com/greenplum-db/gpdb/blob/master/gpMgmt/bin/gpcheckcat#L440), which seems to return all primaries and both the master and master standby.

The function needs to be updated with another condition to check for 'isprimary' is true. Something like `master_segment, = filter(lambda config: config['content'] == -1 and config['isprimary'], self._context.cfg.values())`

File: gpMgmt/bin/gpcheckcat [hunk 7]
```diff
@@ -2908,6 +2909,51 @@ def log_unique_index_violations(violations):
...
+        logger.info('[OK]  %s' % checkname)
+    else:
+        logger.info('[FAIL]  %s' % checkname)
+        GV.checkStatus = False
+        setError(ERROR_REMOVE)
+
+        log_output = ['Orphan toast tables detected for the following scenarios:']
+        for issue, rows in checker.get_issue_to_rows().iteritems():
+            log_output.append('\n' + issue.header)
+            log_output.append('content_id | toast_table_oid | toast_table_name | expected_table_oid | expected_table_name | dependent_table_oid | dependent_table_name')
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat]: For my own edification, can you point me to where this code is? Maybe I can leverage it in this PR?

[the remark this answers is not in the document]
Comment (jchampio) [reply gpMgmt/bin/gpcheckcat]: `gppylib.utils.TableLogger` is the class I am thinking of. I wouldn't be surprised to find that it needs improvement before it can be used here.

**Comment by berlin-ab** [comment]
What's the status of this PR?

File: gpMgmt/bin/gpcheckcat (hunk 4 above)
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat]: To make the function easier to read, we added comments about each of the different logging. I don't think it makes sense to split the function since the logging functions cannot be reused, and the function fits on one screen.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py (hunk 2 above)
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Yes, these have now been removed.

**Comment by kalensk** [comment]
> For some of the commit messages, a line length maximum would be good. The 50/72 rule is one I like -- it's not important to follow it religiously, but we should have something other than "infinity" columns per line.

Fixed!

File: gpMgmt/bin/gpcheckcat_modules/repair.py [hunk 8]
```diff
@@ -25,26 +25,36 @@ def __init__(self, context=None, issue_type=None, desc=None):
...
         extra_missing_repair_obj = RepairMissingExtraneous(catalog_table_obj=catalog_table_obj,
                                                            issues=issues,
                                                            pk_name=pk_name)
         repair_dir = self.create_repair_dir()
-        all_seg_ids = self._context.report_cfg.keys()
-        segment_to_oid_map = extra_missing_repair_obj.get_segment_to_oid_mapping(all_seg_ids)
+        segment_to_oids_map = extra_missing_repair_obj.get_segment_to_oid_mapping(map(lambda config: config['content'], segments.values()))
+
+        for segment_id, oids in segment_to_oids_map.iteritems():
+            segment, = filter(lambda config: config['content'] == segment_id, segments.values())
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: Updated to use python list comprehension syntax, such as:

`master_segment = next(segment for segment in self._context.cfg.values() if segment['content'] == -1)`

This is not quite ideal since we need to extract the first element from the returned list. Ideas?

File: gpMgmt/bin/gpcheckcat_modules/repair.py [hunk 9]
```diff
@@ -65,65 +75,68 @@ def append_content_to_bash_script(self, repair_dir_path, script, catalog_name=No
...
         bash_script_content = '\necho "{0}"\n'.format(self._desc)
-        bash_script_content += self.__get_psql_command(segment_id).format(hostname=c['hostname'],
-                                                                          port=c['port'], sql=filename,
-                                                                          dbname=self._context.dbname,
-                                                                          filename=out_filename)
+        bash_script_content += self.__get_psql_command(segment, sql_filename, out_filename)
         return bash_script_content
 
-    def __get_psql_command(self, segment_id):
+    def __get_psql_command(self, segment, sql_filename, out_filename):
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: I updated the various variable names to be more accurate. Hopefully, its a bit better now without doing a huge overhaul.

File: gpMgmt/bin/gpcheckcat (hunk 7 above)
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat]: I experimented with `gppylib.utils.TableLogger` and it does need some work to look nice. I will definitely keep it in mind for the future!

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    from gppylib.db import dbconn
+
+except ImportError, e:
+    sys.exit('Error: unable to import module: ' + str(e))
+
+class OrphanedToastTablesCheck:
+
+
+    def __init__(self):
+        self.table_to_issues = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))  # a dictionary of orphan tables to a dictionary of issue types to list of content_ids
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: New implementation using `OrphanToastTableIssue` class and multiple data structures. It feels a lot easier to reason about now.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    def get_table_to_issues(self):
+        return self.table_to_issues
+
+    def get_issue_to_rows(self):
+        return self.issues_to_rows
+
+    def get_fix_text(self):
+        log_output = ['\nORPHAN TOAST TABLE FIXES:', '===================================================================']
+        for issue in self.issues_to_rows.keys():
+            if self.issues_to_rows[issue] and issue.type == 'double_orphans':
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Should be cleaner with the introduction of `OrphanToastTableIssue` class.

File: gpMgmt/test/behave/mgmt_utils/gpcheckcat.feature [hunk 12]
```diff
@@ -426,3 +426,110 @@ Feature: gpcheckcat tests
...
+        And run all the repair scripts in the dir "repair_dir"
+        When the user runs "gpcheckcat -R orphaned_toast_tables -g repair_dir gpcheckcat_orphans"
+        And gpcheckcat should print "catalog issue\(s\) found , repair script\(s\) generated" to stdout
+        And the user runs "dropdb gpcheckcat_orphans"
+        And the path "repair_dir" is removed from current working directory
+
+    @orphaned_toast
+    Scenario: gpcheckcat should log and not repair "mismatched non-cyclic" orphaned toast tables (caused by non-matching reltoastrelid)
+        Given the database "gpcheckcat_orphans" is initialized for orphaned TOAST table tests
+        And the database "gpcheckcat_orphans" is broken with "mismatched non-cyclic" orphaned toast tables
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/gpcheckcat.feature]: Yes, I know what you mean. Unfortunately, Im not quite sure how to make it more explicit while keep things "clean". For what its worth, we do describe what the actual problem means (ie: "mismatched non-cyclic") in the `Scenario` line (ie: caused by non-matching reltoastrelid).

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+            if row['expected_table_oid'] is None and row['dependent_table_oid'] is None:
+                issue = Issue('double_orphans',
+                              'Double Orphan TOAST tables due to a missing reltoastrelid in pg_class and a missing pg_depend entry.',
+                              'Found a "double orphan" orphaned TOAST table caused by missing reltoastrelid and missing pg_depend entry.')
+                issue_cause = 'is an orphan TOAST table.'
+                orphan_table = OrphanedTable(row['toast_table_oid'], row['toast_table_name'])
+
+                self.issues_to_rows[issue].append(row)
+                self.table_to_issues[orphan_table][issue][issue_cause].append(row['content_id'])
```
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: New implementation using `OrphanToastTableIssue` class and multiple data structures. It feels a lot easier to reason about now.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py (hunk 3 above)
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Should be a bit easier to reason about with the new implementation using `OrphanToastTableIssue` class and multiple data structures.

**Comment by jmcatamney** [comment]
The description for this PR, and the commit message for 6c9784626cb2bbee716bc1c01e79a21fb99b8b83, mentions the "Bad Reference" orphan type twice.  The second bullet should be changed to the following:

* Bad Dependency Orphaned TOAST tables due to a missing entry in pg_depend.

**Comment by kalensk** [comment]
> The description for this PR, and the commit message for [6c97846](https://github.com/greenplum-db/gpdb/commit/6c9784626cb2bbee716bc1c01e79a21fb99b8b83), mentions the "Bad Reference" orphan type twice. The second bullet should be changed to the following:
> 
> * Bad Dependency Orphaned TOAST tables due to a missing entry in pg_depend.
> 
> Otherwise, looks good to me.

Fixed. Thanks.

File: gpMgmt/bin/gpcheckcat_modules/repair.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jchampio) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: I took a deeper look and finally figured out what was bothering me -- the `GV.cfg` maps dbid to segment instead of content ID to segment. It looks like there are many uses of `GV.cfg[1]` throughout gpcheckcat to mean the master segment (which is incorrect). Not a job for this PR, but once we fix that, this should be as easy as `GV.cfg[-1]`.

**Comment by jchampio** [comment]
@kalensk I have one last concern:
> Basically, if we can't find the valid reloid then the repair script does not work. And the user friendly text for the fixes states `If the dependent table is invalid, the associated TOAST table has been renamed. A manual catalog change is needed.`

Is there logic to make sure that the toast table that we've looked up by name doesn't actually belong to another table? E.g. table 10000 could orphan its TOAST table named `pg_toast_9500`, while `pg_toast_10000` legitimately belongs to table 15000.

File: gpMgmt/bin/gpcheckcat_modules/repair.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kalensk) [reply gpMgmt/bin/gpcheckcat_modules/repair.py]: Yup, I had the same thought.

**Comment by jchampio** [comment]
Rebased on top of latest master.

@kalensk @tvesely Jimmy and I added some tests for an additional corner case: if a single table has multiple types of "orphan issues" on different segments. Getting the model to handle that case led to some further refactoring and simplification of the orphan checker API. All those changes are in the last commit that we pushed.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 14]
```diff
@@ -122,63 +118,84 @@ def runCheck(self, db_connection):
...
+            return (issue.table.oid, type(issue))
+
+        def key_sort(this, that):
+            return cmp(hash(issue_key(this)), hash(issue_key(that)))
+
+        sorted_issues = sorted(self._issues, key_sort)
+
+        for _, group in itertools.groupby(sorted_issues, issue_key):
+            issues = list(group)
+            yield issues[0], { i.row['content_id'] for i in issues }
```
Comment (tvesely) [inline gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: It's kind of hard to follow this bit of code, you need to read the entire comment above to understand what it is returning. It would be nice if you could just give it a 'cheat sheet' as a comment above the `yield` for people that want to understand what you are returning, but don't necessarily need to understand the _how_ or _why_.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py [hunk 15]
```diff
@@ -1,22 +1,18 @@
...
 
 OrphanedTable = namedtuple('OrphanedTable', 'oid catname')
 
 
 class OrphanedToastTablesCheck:
     def __init__(self):
-        self.table_to_issue = defaultdict()  # a dictionary of orphan table to issue (header, description, cause)
-        self.table_to_segments = defaultdict(list)  # a dictionary of orphan table to affected segments (content_id's)
-        self.issue_to_rows = defaultdict(list)  # a dictionary of issue to individual orphan toast tables
+        self._issues = []
```
Comment (kalensk) [inline gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Its nice that there is one data structure with different ways of viewing the data. Even though I find having two data structures easier to follow than the `iterate_issues` approach. I suppose its somewhat personal preference.

I did learn a bit more about python with the `iterate_issues` approach. Thanks.

**Comment by kalensk** [comment]
The commit for `gpcheckcat: add check for orphaned TOAST tables` mentions not being able to generate a repair script for msimatch orphans. We now also do not generate a repair script for double orphans.

Looks good overall though.

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py (hunk 15 above)
Comment (jchampio) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: Yeah, this one was hard. I'm still pretty convinced that this is not what we want long-term; the `itertools` approach is non-obvious.

> Even though I find having two data structures easier to follow than the `iterate_issues` approach.

In this case, having multiple models made it more difficult to figure out which one was the "correct" one (it turned out the `table_to_issue` model was buggy, in that a single table can have multiple issues). IMO, a buggy view is generally easier to untangle than having multiple sources of truth telling you slightly different things.

**Comment by jchampio** [comment]
> The commit for `gpcheckcat: add check for orphaned TOAST tables` mentions not being able to generate a repair script for msimatch orphans. We now also do not generate a repair script for double orphans.

Added a note in that commit message, thanks!

File: gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py (hunk 14 above)
Comment (jchampio) [reply gpMgmt/bin/gpcheckcat_modules/orphaned_toast_tables_check.py]: I'm having trouble coming up with a good comment that doesn't just reiterate what's already in the `yield` statement. I'll push this since it looks like we're all comfortable with the implementation, and keep thinking about a better way to document...
